File: LinkedList/Singly/linked_list.c
#include<stdio.h>
#include<stdlib.h>

typedef struct Node {
	int data;
	struct Node* next;
} Node;

int length = 0;

Node* create_node(int data){
	Node* new_node = malloc(sizeof(Node));
	new_node -> data = data;
	return new_node;
}

void insert(Node** head, int value){
	Node* new_node = create_node(value);
	new_node->next = *head;
	*head = new_node;
	length++;
}

void delete(Node** head){
	if(*head == NULL) {
		printf("-=-=-=-=- Trying to delete NULL lists -=-=-=-=-\n");
	}
	Node* node_to_delete = *head;
	*head = node_to_delete -> next;
	free(node_to_delete);
	length--;
}

void insert_at(Node** head, int value, int index){
	if(index < 0 || index > length){
		printf("-=-=-=-=- Trying to add value at overlapping index -=-=-=-=-\n");
		return;
	}

	length++;
	Node* new_node = create_node(value);
	if(index == 0){
		new_node -> next = *head;
		*head = new_node;
		return;
	}

	Node* node = *head;
	for(int i = 0; i < index - 1; i++){
		node = node -> next;
	}

	new_node -> next = node -> next;
	node -> next = new_node;
}

void delete_at(Node** head, int index){
	if(index < 0 || index >= length || *head == NULL){
		printf("-=-=-=-=- Trying to remove value at overlapping index or an empty list -=-=-=-=-\n");
		return;
	}

	Node* node = *head;
	length--;
	if(index == 0){
		*head = node -> next;
		free(node);
		return;
	}

	for(int i = 0; i < index - 1; i++){
		node = node->next;
	}

	Node* node_to_delete = node->next;
	node->next = node_to_delete->next;
	free(node_to_delete);
}

Node* reversed_recurse(Node** head, Node* node){
	if(*head == NULL) return NULL;
	if(node -> next == NULL){
		*head = node;
		return node;
	}
	Node* next_node = reversed_recurse(head, node->next);
	node->next = NULL;
	next_node->next = node;
	return node;
}

void reversed_iterative(Node** head){
	if(length <= 1 || *head == NULL){
		printf("-=-=-=-=- list empty or has only one node -=-=-=-=-\n");
		return;
	}

	Node* prev = NULL;
	Node* current = *head;
	Node* next = current->next;

	while(current->next != NULL){
		current->next = prev;
		prev = current;
		current = next;
		next = current->next;
	}

	current->next = prev;
	*head = current;
}

void traverse(Node* head){
	Node* node = head;
	printf(" --- Traversing Linked list ---\n");
	while(node != NULL){
		printf("%d -> ", node->data);
		node = node -> next;
	}
	printf(" NULL\n");
}

void traverse_recurse(Node* head){
	if (head == NULL) return;
	traverse_recurse(head->next);
	printf("%d -> ", head->data);
}

void free_all(Node* head){
	Node* node = head;
	while(node != NULL){
		Node* node_to_delete = node;
		node = node -> next;
		free(node_to_delete);
	}
}

int main(){
	Node* head = NULL;
	insert(&head, 77);
	insert(&head, 5);
	insert(&head, 9);
	insert(&head, 10);
	insert(&head, 56);
	
	traverse(head);
	//delete(&head);
	// printf("List's length = %d\n", length);
	// insert_at(&head, 69, 2);
	// printf("List's length = %d\n", length);
	
	delete_at(&head, 2);
	traverse(head);
	reversed_iterative(&head);
	traverse(head);
	reversed_recurse(&head, head);
	traverse(head);
	traverse_recurse(head); printf(" NULL\n");

	free_all(head);
}

File: LinkedList/Singly/linked_list.py
class Node:
    def __init__(self, data = None):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None
        self.length = 0

    def insert(self, value: int):
        new_node = Node(value)
        new_node.next = self.head
        self.head = new_node
        self.length += 1

    def insert_at(self, value: int, index: int):
        if index > self.length or index < 0:
            print("Trying to add value to overlapping index")
            return
        new_node = Node(value)
        self.length += 1
        if index == 0:
            new_node.next = self.head
            self.head = new_node
            return

        node = self.head
        for i in range(index - 1):
            node = node.next

        new_node.next = node.next
        node.next = new_node

    
    def remove(self):
        if self.length <= 0 or self.head == None:
            print("Trying to remove to an empty list")
            return
        node_to_remove = self.head
        self.head = node_to_remove.next
        node_to_remove = None
        self.length -= 1
    
    def remove_at(self, index: int):
        if index > self.length -1 or index < 0:
            print("Trying to remove value to overlapping index")
            return
        
        node = self.head
        self.length -= 1
        if index == 0:
            self.head = node.next
            node = None
            return

        for i in range(index - 1):
            node = node.next

        node_to_delete = node.next
        node.next = node_to_delete.next
        node_to_delete = None

    def traverse(self):
        print("--- TRAVERSE LINKED LIST ---")
        node = self.head
        while node != None:
            print(f"{node.data} -> ", end = " ")
            node = node.next

        print("null")

    def reverse_traverse(self, node: Node):
        if node == None:
            return
        self.reverse_traverse(node.next)
        print(f"{node.data} -> ", end = " ")

    def reversed_recurse(self, node: Node):
        if node == None:
            return
        if node.next == None:
            self.head = node
            return node

        next_node = self.reversed_recurse(node.next)
        next_node.next = node
        node.next = None
        return node
    
    def reversed_iterative(self):
        if self.length <= 1 or self.head == None:
            print("Empty list or list has only one node.")
            return

        prev = None
        current = self.head
        next_node = current.next
        
        while(current.next != None):
            current.next = prev
            prev = current
            current = next_node
            next_node = current.next

        current.next = prev
        self.head = current


node = LinkedList()
node.insert(77)
node.insert(5)
node.insert(9)
node.insert(10)
node.insert(56)

node.traverse()
node.reverse_traverse(node.head)
print()

# node.remove()
# node.remove()

node.insert_at(69, 6)
node.traverse()
print(node.length)

node.remove_at(5)
node.traverse()
print(node.length)

node.reversed_recurse(node.head)
node.traverse()

node.reversed_iterative()
node.traverse()

print(node.length)
